Marlin/src/HAL/AVR/fast_pwm.cpp

```diff
@@ -264,8 +264,13 @@ void set_pwm_duty(const pin_t pin, const uint16_t v, const uint16_t v_size/*=255
         ), COM_CLEAR_SET + invert
       );
 
-      uint16_t top = (timer.n == 2) ? TERN(USE_OCR2A_AS_TOP, *timer.OCRnQ[0], 255) : *timer.ICRn;
-      _SET_OCRnQ(timer.OCRnQ, timer.q, (v * top + v_size / 2) / v_size); // Scale 8/16-bit v to top value
+      uint16_t top;
+      if (timer.n == 2)     // if TIMER2
+        top = TERN(USE_OCR2A_AS_TOP, *timer.OCRnQ[0], 255);
+      else {
+        top = *timer.ICRn;  // top = ICRn
+        _SET_OCRnQ(timer.OCRnQ, timer.q, v * float(top) / float(v_size)); // Scale 8/16-bit v to top value
```

Comment:
Note that the updated maths gives a result of `floor(n)` in contrast to the previous result of `round(n)`.